scripts/build-ancillary-repository.sh

```diff
@@ -119,23 +162,45 @@ AWS_S3_URI_COMBINED_PACKAGES=$(resolve_s3_uri \
 	"$AWS_S3_URI_COMBINED_PACKAGES" "" \
 	"devops-gate/master/linux-pkg/${UPSTREAM_BRANCH}/combine-packages/post-push/latest")
 
+mkdir -p "$TOP/build"
+WORK_DIRECTORY=$(mktemp -d -p "$TOP/build" tmp.pkgs.XXXXXXXXXX)
+
 #
 # All package files will be placed into this temporary directory, such
 # that we can later point Aptly at this directory to build the Aptly/APT
 # repository.
 #
-mkdir -p "$TOP/build"
-PKG_DIRECTORY=$(mktemp -d -p "$TOP/build" tmp.pkgs.XXXXXXXXXX)
+PKG_DIRECTORY="$WORK_DIRECTORY/debs/"
+mkdir -p "$PKG_DIRECTORY"
+
+#
+# Download all the packages and the associated metadata from the
+# combined-packages S3 location.
+#
+aws s3 sync --only-show-errors "$AWS_S3_URI_COMBINED_PACKAGES" "$WORK_DIRECTORY"
+
+#
+# Prepare a build-info file while we have access to the downloaded build
+# metadata files. Later in the build this file will be included in the
+# delphix-entire package.
+#
+compile_build_info "$WORK_DIRECTORY" >"$BUILD_INFO_FILE"
```

Comment:
Why are we generating this file at this point in the build?

At first glance, this doesn't really seem directly related to building the ancillary repository, so it just seems a little awkward that you're choosing to generate the file here.

Further, this file is only consumed by the delphix-entire package, so if we just look at this script in isolation, it seems super awkward to generate it and not use it for anything.

For example, if we want to run this script manually (or any other context other than how we currently call it) to build that ancillary repository, would we also need/want to generate this file? In such a case, this seems like it may be an undesirable side-effect.

I'm not sure how best to implement this, but I feel like a better solution would be to have a separate script, whose sole responsibility was to generate this build information. And then have an explicit place in the build, e.g. when we build delphix-entire, where the script was called to generate the file.

Along the same lines, I think it'd be nice to generate this file closer to the logic that would consume the file. As is, we generate the file, and then have to jump through some non-intuitive hoops to ensure this file is properly propagated through the build to the point where it's consumed (i.e. the change to `run-live-build.sh`).

Reply:
The main reason why I'm generating this here is that it's the only time that we have access to downloaded packages.

I agree that it is a bit awkward to put this into build-ancillary-repository. I can look into creating a new function that would download the metadata and generate this info when generating the delphix-entire package (or even create a separate package, although I don't feel like that's needed).